### ISSUE TITLE
Translate text emojis to Unicode characters

### ISSUE THREAD

**Issue by tgy** [matterhorn-chat/matterhorn#597]
![image](https://user-images.githubusercontent.com/694873/79049928-d7d6c700-7c26-11ea-872d-b3a7b3b0e3d6.png)

When receiving emojis, they are displayed in my terminal.

However, when I sent them, they seem to not show up as emojis.

Is there a reason for this? I would expect the same behaviour for sent and received messages.

```
( ⌁ ) matterhorn --version
matterhorn 50200.6.0 (master@789a293)
 using mattermost-api 50200.3.0 (master@661e7d4)
```

**Comment by jtdaugherty** [comment]
Thanks for the report!

The reason is that the other user's client sends the Unicode character (the heart in this case) whereas Matterhorn does no such translation. This is partly because we haven't put in the work to do that translation -- the data we have available to translate the text versions into Unicode isn't in a great format -- but also because, frankly, we haven't made it a priority due to the lack of good support for Unicode characters in terminals. [Up until recently](https://github.com/matterhorn-chat/matterhorn/issues/569) there was no good way for Matterhorn to reliably deal with these characters because of terminal technology issues, which meant that even if we had translated to the Unicode characters, they likely would have resulted in rendering problems in the Matterhorn UI (which was also true for Unicode emoji characters coming from other clients).

I think that after [the Unicode width table ticket](https://github.com/matterhorn-chat/matterhorn/issues/569) has been addressed, we'll be in a better position to do the translation needed to address this ticket. And when we do, I'd like it to be configurable and default to "off", i.e., the current behavior, due to the risk of rendering problems if a width table is not installed.

In short, there is a bit of history behind supporting this feature and I'm happy to provide more pointers and context if you're curious. In the mean time we can have this ticket as a reminder that once width table support is ready, there's more we can do.

**Comment by jtdaugherty** [comment]
Another comment for us to consider when we come back around to this: the web client renders incoming text emoji using the Unicode equivalent. That means that if I send a message containing the text `:+1:`, the web client will render that as the Unicode emoji version. But the web client *sends* Unicode emoji characters: it does the replacement before sending. I think that's a bad practice because it breaks the separation of concerns for presentation of emoji. In Matterhorn I'd like us to just support rendering, but never *send* Unicode emoji -- always send text versions. The other clients can then decide how to render them. In practical terms, this means just doing the text-to-Unicode-glyph translation at rendering time when we convert Markdown ASTs to Brick renderings.

**Comment by tgy** [comment]
Thank you for this lengthy reply 👍

**Comment by jtdaugherty** [comment]
This ticket has been labeled `shelved` due to lack of time and/or urgency to work on this, although it may still be worthwhile to address in the long run. Anyone wanting to contribute by working on this ticket can re-open.

**Comment by 3v1n0** [comment]
Oh, sad to hear... This is actually my mostly awaited issue :stuck_out_tongue:

**Comment by jtdaugherty** [comment]
Sorry, @3v1n0 - this particular issue also depends on some other issues getting resolved, particularly #569, which has turned out to be more difficult to do than we anticipated. :(
